src/System.Private.ServiceModel/src/System/ServiceModel/Dispatcher/ImmutableDispatchRuntime.cs

```diff
@@ -0,0 +1,904 @@
+                        {
+                            throw;
+                        }
+                        thereIsAnUnhandledException = !_error.HandleError(e);
+                        exception = e;
+                    }
+                }
+            }
+
+            if (rpc.Operation.IsOneWay)
+            {
+                rpc.CanSendReply = false;
+            }
+
+            if (!rpc.Operation.IsOneWay && (context != null) && (rpc.Reply != null))
+            {
+                if (exception != null)
+                {
+                    // We don't call ProvideFault again, since we have already passed the
+                    // point where SFx addresses the reply, and it is reasonable for
+                    // ProvideFault to expect that SFx will address the reply.  Instead
+                    // we always just do 'internal server error' processing.
+                    rpc.Error = exception;
+                    _error.ProvideOnlyFaultOfLastResort(ref rpc);
+
+                    try
+                    {
+                        rpc.CanSendReply = PrepareAndAddressReply(ref rpc);
+                    }
+                    catch (Exception e)
+                    {
+                        if (Fx.IsFatal(e))
+                        {
+                            throw;
+                        }
+                        _error.HandleError(e);
+                    }
+                }
+            }
+            else if ((exception != null) && thereIsAnUnhandledException)
+            {
+                rpc.Abort();
+            }
+        }
+
+        bool PrepareAndAddressReply(ref MessageRpc rpc)
+        {
+            bool canSendReply = true;
+
+            if (!_manualAddressing)
+            {
+                if (!object.ReferenceEquals(rpc.RequestID, null))
+                {
+                    System.ServiceModel.Channels.RequestReplyCorrelator.PrepareReply(rpc.Reply, rpc.RequestID);
+                }
+
+                if (!rpc.Channel.HasSession)
+                {
+                    canSendReply = System.ServiceModel.Channels.RequestReplyCorrelator.AddressReply(rpc.Reply, rpc.ReplyToInfo);
+                }
+            }
+
+            AddMessageProperties(rpc.Reply, rpc.OperationContext, rpc.Channel);
+            if (FxTrace.Trace.IsEnd2EndActivityTracingEnabled && rpc.EventTraceActivity != null)
+            {
+                rpc.Reply.Properties[EventTraceActivity.Name] = rpc.EventTraceActivity;
+            }
+
+            return canSendReply;
+        }
+
+        internal DispatchOperationRuntime GetOperation(ref Message message)
+        {
+            return _demuxer.GetOperation(ref message);
+        }
+
+        interface IDemuxer
+        {
+            DispatchOperationRuntime GetOperation(ref Message request);
+        }
+
+        internal bool IsConcurrent(ref MessageRpc rpc)
+        {
+            return _concurrency.IsConcurrent(ref rpc);
+        }
+
```

Comment:
Good suggestions but I recommend opening a new issue for cleanup and not complicate this PR any further.